README.md
# Mercado Pago Helper

[![pub package](https://img.shields.io/pub/v/mercado_pago.svg)](https://pub.dartlang.org/packages/mercado_pago)

A flutter plugin to integrate MercadoPago into flutter app. So far I completed the basic functions which includes:
- New user
- New card
- Associate card with user
- Get all card by user
- Token with card
- Create payment

## Usage

Setup MercadoPago credentials first
```dart
import 'package:mercado_pago/mercado_pago.dart';
final credentials = MercadoCredentials(
    publicKey: 'PUBLIC-KEY-TEST',
    accessToken: 'ACCESS-TOKE-TEST'
);
```

> This plugin is in very initial stage, I am using it in my personal app. Future plan is to complete all the remaining API. 

CHANGELOG.md
## [0.0.1] - 16/12/2018

* Create new user
* Create new card
* TODO: associate card with user

## [1.0.5] - 27/04/2020

* Fix typeInt
* add unit test
* Get document types

## [1.0.6] - 27/04/2020

* Format code


test/api_test.dart
import 'package:test/test.dart';
import 'package:mercado_pago/mercado_pago.dart';

MercadoPago get mercadoPago {
  List<String> keys = [
    'TEST-98041829-8c47-4c6a-9c23-7b6e1855f31d',
    'TEST-3029117202042245-103104-2fd0688859e43720378e5ed1043114f4__LC_LB__-182447115',
  ];

  MercadoCredentials credentials = MercadoCredentials(
    publicKey: keys[0],
    accessToken: keys[1],
  );

  return MercadoPago(credentials);
}

void main() {
  test('get document types', () async {
    MercadoObject response = await mercadoPago.documentTypes();
    print(response);
  });

  test('create new user', () async {
    MercadoObject response = await mercadoPago.newUser(
      email: 'brian1@mail.com',
      firstname: 'Brian',
      lastName: 'Castillo',
      documentType: 'DNI',
      documentNumber: '12345678',
    );
    print(response);
  });

  test('get user', () async {
    String userId = '555305508-i67KHqcUTewosJ';
    MercadoObject response = await mercadoPago.getUser(userId);
    print(response);
  });

  test('create new card', () async {
    MercadoObject response = await mercadoPago.newCard(
      code: '333',
      year: '2020',
      month: 9,
      card: '4009175332806176',
      documentNumber: '12345678',
      documentType: 'DNI',
      fullName: 'APRO',
    );
    print(response);
  });

  test('associate card with user', () async {
    String cardId = 'ee6bbbee69f60990d0f68ffe108ef1ad';
    String userId = '555305508-i67KHqcUTewosJ';
    MercadoObject response = await mercadoPago.associateCardWithUser(
      user: userId,
      card: cardId,
    );
    print(response);
  });

  test('get cards for user', () async {
    String userId = '555305508-i67KHqcUTewosJ';
    MercadoObject response = await mercadoPago.cardsFromUser(
      user: userId,
    );
    print(response);
  });

  test('create card token for payment', () async {
    String cardId = '1587964933876';
    String cardCVV = '333';
    MercadoObject response = await mercadoPago.tokenWithCard(
      card: cardId,
      code: cardCVV,
    );
    print(response);
  });

  test('simple payment', () async {
    String cardToken = 'ebcc4d445e845f052f702ed7015c4d57';
    String userId = '555305508-i67KHqcUTewosJ';
    MercadoObject response = await mercadoPago.createPayment(
      total: 10.0,
      cardToken: cardToken,
      description: 'test payment',
      paymentMethod: 'visa',
      userId: userId,
      email: 'brian1@mail.com',
    );
    print(response);
  });
}


test/widget_test.dart
import 'package:mercado_pago/mercado_pago.dart';

void main() {

}


lib/mercado_page.dart
library mercado_pago;

import 'package:http/http.dart' as http;

import 'dart:convert';
import 'dart:async';

class MercadoPago {
  static const _base_url = 'https://api.mercadopago.com';

  final MercadoCredentials _mercadoCredentials;

  String get _publicKey => _mercadoCredentials.publicKey;
  String get _accessToken => _mercadoCredentials.accessToken;

  MercadoPago(this._mercadoCredentials);

  Future<MercadoObject> newUser({
    String firstname, 
    String lastName, 
    String email
  }) async {
    final url = '$_base_url/v1/customers?access_token=$_accessToken';

    var body = {
      'email': email,
      'first_name': firstname,
      'last_name': lastName,
    };

    var response = await http.post(url, body: json.encode(body));

    var jsonBody = json.decode(response.body);

    MercadoObject responseObject = MercadoObject();

    if (response.statusCode == 201) {
      responseObject.isSuccessful = true;
      responseObject.data = {'id': jsonBody['id']};
    } else {
      int errorCode = int.tryParse(jsonBody["cause"][0]["code"]) ?? 404;

      if (errorCode == 101) {
        errorCode = 409;
      }

      responseObject.isSuccessful = false;
      responseObject.errorCode = errorCode;
    }

    return responseObject;
  }

  Future<MercadoObject> newCard({
    String code,
    String year,
    int month,
    String card,
    String docNumber,
    String docType,
  }) async {
    final url = '$_base_url/v1/card_tokens?public_key=$_publicKey';

    var body = {
      'security_code': code,
      'expiration_year': year,
      'expiration_month': month,
      'card_number': card,
      'cardholder': {
        'identification': {'number': docNumber, 'type': docType},
        'name': 'WIBO-CARD'
      }
    };

    var response = await http.post(url, body: json.encode(body));

    var jsonBody = json.decode(response.body);

    MercadoObject responseObject = MercadoObject();

    if (response.statusCode == 201) {
      responseObject.isSuccessful = true;
      responseObject.data = {
        'id': jsonBody['id'],
        'key': jsonBody['public_key']
      };
    } else {
      responseObject.isSuccessful = false;
      responseObject.errorCode = 404;
    }

    return responseObject;
  }
}

class MercadoCredentials {
  String publicKey;
  String accessToken;

  MercadoCredentials({this.publicKey, this.accessToken});
}

class MercadoObject {
  bool isSuccessful;
  Map data;
  int errorCode;

  MercadoObject({this.isSuccessful, this.data, this.errorCode});

  @override
  String toString() => data.toString();
}


lib/src/services.dart
// Copyright (c) 2018, codegrue. All rights reserved. Use of this source code
// is governed by the MIT license that can be found in the LICENSE file.
import 'dart:convert';
import 'dart:async';

import 'package:flutter/foundation.dart' show required;
import 'package:http/http.dart' as http;

import 'credentials.dart';
import 'model.dart';

// Core of library
class MercadoPago {
  // base url
  static const String _base_url = 'https://api.mercadopago.com';

  final MercadoCredentials _mercadoCredentials;

  // getters of credentials
  String get _publicKey => _mercadoCredentials.publicKey;
  String get _accessToken => _mercadoCredentials.accessToken;

  MercadoPago(this._mercadoCredentials);

  // same process of response
  Future<MercadoObject> _response(
    dynamic response, {
    int customSuccessCode = 201,
  }) async {
    /// decode response
    var jsonBody = json.decode(response.body);

    MercadoObject responseObject = MercadoObject();
    responseObject.isSuccessful = true;
    responseObject.data = jsonBody;

    /// if error
    if (response.statusCode != customSuccessCode) {
      responseObject.isSuccessful = false;

      try {
        responseObject.errorCode = jsonBody["cause"][0]["code"];
      } //
      catch (_) {
        responseObject.errorCode = response.statusCode.toString();
      }
    }

    return responseObject;
  }

  // http get without body
  Future<MercadoObject> _get(
    String url, {
    int customSuccessCode = 201,
  }) async {
    var response = await http.get(url);

    return _response(response, customSuccessCode: customSuccessCode);
  }

  // http post with body
  Future<MercadoObject> _post(
    String url,
    dynamic body, {
    int customSuccessCode = 201,
  }) async {
    var response = await http.post(url, body: json.encode(body));

    return _response(response, customSuccessCode: customSuccessCode);
  }

  Future<MercadoObject> documentTypes() async {
    final url = '$_base_url/v1/identification_types?access_token=$_accessToken';

    return await _get(url, customSuccessCode: 200);
  }

  Future<MercadoObject> newUser({
    @required String firstname,
    @required String lastName,
    @required String email,
    @required String documentType,
    @required String documentNumber,
  }) async {
    final url = '$_base_url/v1/customers?access_token=$_accessToken';

    var body = {
      'email': email,
      'first_name': firstname,
      'last_name': lastName,
      'identification': {
        'type': documentType,
        'number': documentNumber,
      },
    };

    return await _post(url, body);
  }

  Future<MercadoObject> getUser(String userId) async {
    final url = '$_base_url/v1/customers/$userId?access_token=$_accessToken';

    return await _get(url);
  }

  Future<MercadoObject> newCard({
    @required String code,
    @required String year,
    @required int month,
    @required String card,
    @required String documentType,
    @required String documentNumber,
    @required String fullName,
  }) async {
    final url = '$_base_url/v1/card_tokens?public_key=$_publicKey';

    var body = {
      'security_code': code,
      'expiration_year': year,
      'expiration_month': month,
      'card_number': card,
      'cardholder': {
        'identification': {
          'number': documentNumber,
          'type': documentType,
        },
        'name': fullName
      }
    };

    return await _post(url, body);
  }

  Future<MercadoObject> associateCardWithUser({
    @required String user,
    @required String card,
  }) async {
    final url =
        '$_base_url/v1/customers/$user/cards?access_token=$_accessToken';

    var body = {'token': card};

    return await _post(url, body);
  }

  Future<MercadoObject> cardsFromUser({
    @required String user,
  }) async {
    final url =
        '$_base_url/v1/customers/$user/cards?access_token=$_accessToken';

    return await _get(url, customSuccessCode: 200);
  }

  Future<MercadoObject> tokenWithCard({
    @required String code,
    @required String card,
  }) async {
    final url = '$_base_url/v1/card_tokens?public_key=$_publicKey';

    var body = {'security_code': code, 'cardId': card};

    return await _post(url, body);
  }

  Future<MercadoObject> createPayment({
    @required double total,
    @required String cardToken,
    @required String description,
    @required String paymentMethod,
    @required String userId,
    @required String email,
  }) async {
    final url = '$_base_url/v1/payments?access_token=$_accessToken';

    var body = {
      'token': cardToken,
      'transaction_amount': total,
      'description': description,
      'installments': 1,
      'payment_method_id': paymentMethod,
      'payer': {
        'id': userId,
        'email': email,
      },
    };

    return await _post(url, body);
  }
}

typedef RewriteResponse = Map Function(dynamic jsonBody);


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:mercado_pago/mercado_pago.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Mercado Pago Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Mercado Pago Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  /*
  * Auth https://www.mercadopago.com/mpe/account/credentials
  */
  final credentials = MercadoCredentials(
      publicKey: 'TEST-98041829-8c47-4c6a-9c23-7b6e1855f31d',
      accessToken:
          'TEST-3029117202042245-103104-2fd0688859e43720378e5ed1043114f4__LC_LB__-182447115');

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            RaisedButton(
              onPressed: () {
                MercadoPago(credentials)
                    .newUser(
                        firstname: 'Brian',
                        lastName: 'Castillo',
                        email: 'brian@castillo90.com')
                    .then((responseObject) {
                    print('user created with id = [ $responseObject ]');
                });
              },
              child: Text('New User'),
            ),
            RaisedButton(
              onPressed: () {
                MercadoPago(credentials)
                    .newCard(
                        code: '333',
                        year: '2020',
                        month: 9,
                        card: '4009175332806176',
                        docNumber: '85695236',
                        docType: 'DNI')
                    .then((responseObject) {
                  print("responseObject => $responseObject");
                });
              },
              child: Text('New Card'),
            ),
            RaisedButton(
              onPressed: () {
                /* MercadoPago(credentials)
                    .associateCardWithUser(
                        user: '387744186-Zw0inDybbSCugR',
                        card: '453c47432aabcf01c7923c6a52060a17')
                    .then((responseObject) {
                  print("responseObject => $responseObject");
                }); */

                MercadoPago(credentials)
                    .cardsFromUser(user: '387744186-Zw0inDybbSCugR')
                    .then((responseObject) {
                  print("cards => ${responseObject.data}");
                });

                /* MercadoPago(credentials)
                    .tokenWithCard(
                      code: '222',
                      card: '1545133239176')
                    .then((responseObject) {
                  print("token => ${responseObject.data}");
                }); */
              },
              child: Text('New Card'),
            ),
            
            RaisedButton(
              onPressed: () {
                MercadoPago(credentials)
                    .createPayment(
                      total: 5.0,
                      cardToken: '8f3e30ddbd518568d6db1ebca76719ae',
                      description: 'Test pay',
                      paymentMethod: 'visa',
                      userId: '387451443-R0Ex7Ij6LdITGi',
                      email: 'jperez1@asasas.com'
                    )
                    .then((responseObject) {
                  print("payment => ${responseObject.data}");
                });
              },
              child: Text('New pay'),
            )
          ],
        ),
      ),
    );
  }
}
